pkg/gardenlet/controller/managedseed/add.go

```diff
@@ -0,0 +1,327 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/gardener/gardener/charts"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes/clientmap"
+	"github.com/gardener/gardener/pkg/controllerutils/mapper"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	confighelper "github.com/gardener/gardener/pkg/gardenlet/apis/config/helper"
+	"github.com/gardener/gardener/pkg/utils"
+	contextutil "github.com/gardener/gardener/pkg/utils/context"
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+const (
+	// ControllerName is the name of this controller.
+	ControllerName = "managedseed"
+
+	// GardenletDefaultKubeconfigSecretName is the default name for the field in the Gardenlet component configuration
+	// .gardenClientConnection.KubeconfigSecret.Name
+	GardenletDefaultKubeconfigSecretName = "gardenlet-kubeconfig"
+	// GardenletDefaultKubeconfigBootstrapSecretName is the default name for the field in the Gardenlet component configuration
+	// .gardenClientConnection.BootstrapKubeconfig.Name
+	GardenletDefaultKubeconfigBootstrapSecretName = "gardenlet-kubeconfig-bootstrap"
+)
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(
+	mgr manager.Manager,
+	cfg config.GardenletConfiguration,
+	gardenCluster cluster.Cluster,
+	seedCluster cluster.Cluster,
+	shootClientMap clientmap.ClientMap,
+	imageVector imagevector.ImageVector,
+) error {
+	if r.GardenClient == nil {
+		r.GardenClient = gardenCluster.GetClient()
+	}
+	if r.GardenNamespaceGarden == "" {
+		r.GardenNamespaceGarden = v1beta1constants.GardenNamespace
+	}
+	if r.GardenNamespaceShoot == "" {
+		r.GardenNamespaceShoot = v1beta1constants.GardenNamespace
+	}
+	if r.ChartsPath == "" {
+		r.ChartsPath = charts.Path
+	}
+
+	valuesHelper := NewValuesHelper(&cfg, imageVector)
+
+	r.Actuator = newActuator(gardenCluster.GetConfig(),
+		gardenCluster.GetAPIReader(),
+		gardenCluster.GetClient(),
+		seedCluster.GetClient(),
+		shootClientMap,
+		valuesHelper,
+		gardenCluster.GetEventRecorderFor(ControllerName+"-controller"),
+		r.ChartsPath,
+		r.GardenNamespaceShoot,
+	)
+
+	// It's not possible to overwrite the event handler when using the controller builder. Hence, we have to build up
+	// the controller manually.
+	c, err := controller.New(
+		ControllerName,
+		mgr,
+		controller.Options{
+			Reconciler:              r,
+			MaxConcurrentReconciles: pointer.IntDeref(r.Config.ConcurrentSyncs, 0),
+			RecoverPanic:            true,
+			RateLimiter:             r.RateLimiter,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	seedName := confighelper.SeedNameFromSeedConfig(cfg.SeedConfig)
```

Comment:
```suggestion
```

I think you don't need to do this. The `cfg.SeedConfig.Name` should always be set. You can probably also remove this helper function.
Instead of this variable, use `cfg.SeedConfig.Name` in L116 and L125.